src/transpile/fusion.hpp

```diff
@@ -628,22 +629,22 @@ bool DiagonalFusion::aggregate_operations(oplist_t& ops,
   // current impl is sensitive to ordering of gates
   for (int op_idx = fusion_start; op_idx < fusion_end; ++op_idx) {
 
+    // find instructions to generate a diagonal gate from op_idx
     std::set<uint_t> checking_qubits_set;
-    auto next_diagonal_end = get_next_diagonal_end(ops, op_idx, checking_qubits_set);
+    auto next_diagonal_end = get_next_diagonal_end(ops, op_idx, fusion_end, checking_qubits_set);
 
     if (next_diagonal_end < 0)
       continue;
 
     if (checking_qubits_set.size() > max_fused_qubits)
       continue;
 
+    // find instructions to generate the next diagonal gates
     auto next_diagonal_start = next_diagonal_end + 1;
 
     while (true) {
-      auto nde = get_next_diagonal_end(ops, next_diagonal_start, checking_qubits_set);
-      if (nde < 0)
-        break;
-      if (checking_qubits_set.size() > max_fused_qubits)
+      auto nde = get_next_diagonal_end(ops, next_diagonal_start, fusion_end, checking_qubits_set);
+      if (nde < 0 || nde >= fusion_end || checking_qubits_set.size() > max_fused_qubits)
```

Comment:
I think in theory `nde` can't ever be greater than `fusion_end`, right?  In `get_next_diagonal_end` it returns `-1` if it reaches the end.  It might be good to write that into the documentation/contract of `get_next_diagonal_end`, and simplify the extraneous condition here.